### PR TITLE
Add support for adding cmake extras to packages

### LINKED ISSUE #331: Feature request: allow packages to append to CMAKE_MODULE_PATH from their cmake config file
## Desired behavior

`gz-cmake3` installs many `cmake` files and adds the containing folder to the `CMAKE_MODULE_PATH` so that packages that `find_package(gz-cmake3)` will be able to use those cmake files.

* https://github.com/gazebosim/gz-cmake/blob/gz-cmake3_3.0.1/config/gz-cmake-config.cmake.in#L61

We would like to install cmake files from other packages as well (such as gz-physics) and append the path of the installed folder to the `CMAKE_MODULE_PATH`. We could expose this functionality as a flag or parameter to the `gz_create_core_library` and `gz_add_component` cmake functions, since that is where the cmake config templates are instantiated.

* `_gz_create_cmake_package function` in [GzPackaging.cmake](https://github.com/gazebosim/gz-cmake/blob/gz-cmake3_3.0.1/cmake/GzPackaging.cmake#L238-L290) is called by [gz_create_core_library](https://github.com/gazebosim/gz-cmake/blob/gz-cmake3_3.0.1/cmake/GzUtils.cmake#L1132) and [gz_add_component](https://github.com/gazebosim/gz-cmake/blob/gz-cmake3_3.0.1/cmake/GzUtils.cmake#L1402) in `GzUtils.cmake`.

## Alternatives considered

Unconditionally add a specific path for each package to the `CMAKE_MODULE_PATH` so it doesn't have to be configured. This may pollute the `CMAKE_MODULE_PATH`, so it would be cleaner to opt in to the functionality.

### PR DESCRIPTION
**Pull request by mjcarroll**
# 🎉 New feature

Closes #331 

## Summary

This addresses #331 in a slightly different way.  Since changing the CMAKE_MODULE_PATH could have strange effects when it comes to resolving cmake files and precedence, this tracks more closely to how `ament_package` allows packages to add cmake functionality.

This adds support for a `CONFIG_EXTRAS` argument to `gz_configure_project`.  In here, users can add cmake files (or templates) that will be sourced by all dependent packages.

I have added a simple demonstration of how this is expected to work here: https://github.com/gazebosim/gz-utils/pull/97

I will have a more detailed example with the message generation workflow later.

## Checklist
- [ ] Signed all commits for DCO
- [ ] Added tests
- [ ] Added example and/or tutorial
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [ ] Consider updating Python bindings (if the library has them)
- [ ] `codecheck` passed (See [contributing](https://gazebosim.org/docs/all/contributing#contributing-code))
- [ ] All tests passed (See [test coverage](https://gazebosim.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Agazebosim+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge** and edit the commit message to match the pull request summary while retaining `Signed-off-by` messages.

### CODE REVIEW

**Comment by azeey** [comment]
The approach seems good to me. The only downside I see is that doing `find_package(gz-cmake)` will implicitly include all the cmake modules configured through `CONFIG_EXTRA` as opposed to each library using `include` to only include the modules they need. Since most libraries won't be using this feature, I don't think we'd need to worry too much about function/macro name clashes. We should probably have a naming convention though.

**Comment by mjcarroll** [comment]
> The only downside I see is that doing find_package(gz-cmake) will implicitly include all the cmake modules configured through CONFIG_EXTRA as opposed to each library using include to only include the modules they need.

`CONFIG_EXTRA` files are scoped to the package that that they are defined in.  `gz-cmake` itself shouldn't make use of this pattern because it installs and includes everything unconditionally.

**Comment by azeey** [comment]
> > The only downside I see is that doing find_package(gz-cmake) will implicitly include all the cmake modules configured through CONFIG_EXTRA as opposed to each library using include to only include the modules they need.
> 
> `CONFIG_EXTRA` files are scoped to the package that that they are defined in. `gz-cmake` itself shouldn't make use of this pattern because it installs and includes everything unconditionally.



So for example, if `gz-physics` added a module using `CONFIG_EXTRA` and I wanted to use it from `gz-chrono`, the module would only be included when I do `find_package(gz-physics)`?

**Comment by mjcarroll** [comment]
> So for example, if gz-physics added a module using CONFIG_EXTRA and I wanted to use it from gz-chrono, the module would only be included when I do find_package(gz-physics)?

This is correct.  Any downstream users will see your module when they did `find_package(gz_chrono)` as well, since `gz-physics` would be a dependency, and the find would be recursive.

I would think that the contents of the config extra should be exporting functions/macros, rather than raw cmake code, which should also help with any namespace pollution.

**Comment by azeey** [comment]
Sounds good! Probably still wise to have a naming convention that would avoid name clashes. One idea is to use the name of the project as a prefix somewhat like [FetchContent](https://cmake.org/cmake/help/latest/module/FetchContent.html).

**Comment by mjcarroll** [comment]
> Probably still wise to have a naming convention that would avoid name clashes

I think it's probably good to have an idea of one, but I wouldn't want to impose it programmatically in case we ever want to do  anything interesting like overriding cmake builtins or overriding gz-cmake functions in downstream packages.

File: cmake/GzConfigureProject.cmake [hunk 1]
```diff
@@ -144,6 +144,56 @@ macro(gz_configure_project)
 
   message(STATUS "${PROJECT_NAME} version ${PROJECT_VERSION_FULL}")
 
+  #============================================================================
+  # Handle extra cmake configurations
+  set(PACKAGE_CONFIG_EXTRA_FILES "")
+  set(extras)
+
+  if (DEFINED PACKAGE_CONFIG_EXTRAS)
```
[the remark this answers is not in the document]
Comment (mjcarroll) [reply cmake/GzConfigureProject.cmake:152]: Hmm, this must have been copied over from `ament_cmake`, but I have no evidence of it being used in the ROS 2 codebase, so we probably don't need it here.

File: cmake/GzUtils.cmake [hunk 2]
```diff
@@ -24,6 +24,60 @@ include(GzInstallAllHeaders)
...
+
+#################################################
+# From ament/ament_cmake_core/core/stamp.cmake (Apache 2.0)
+#   :param path:  file name
+#
+#   Uses ``configure_file`` to generate a file ``filepath.stamp`` hidden
+#   somewhere in the build tree.  This will cause cmake to rebuild its
+#   cache when ``filepath`` is modified.
+#
+function(stamp path)
```
[the remark this answers is not in the document]
Comment (mjcarroll) [reply cmake/GzUtils.cmake]: Done.

**Review by azeey — APPROVED** [review]

**Pull request merged**